### PR TITLE
Fix rationalize for small values

### LINKED ISSUE #16311: Rationalize fails to round to 0//0
I asked on the julia-user list about the `rationalize` function.

> Does rationalize ever return 0//1 for nonzero inputs?
>
> 0.0+eps(Float64) is much closer to 0//1 than to 1//1.
Why can't I get 0//1 out of rationalize(x) if x != 0.0?
Is this to avoid divide by 0 errors when using a/rationalize(x) where 0 < x <
tol?

For example on julia v0.4.5 
```julia
julia> rationalize(0.0)
0//1

julia> rationalize(0.0+eps(Float64))
1//4503599627370496

julia> rationalize(0.0+eps(Float64);tol=1e-4)
1//1

julia> rationalize(0.0+eps(Float64);tol=1e-4)
1//1
```

Jeffrey Sarnoff said:
> This appears to be a bug. The docs for rationalize say 
> rationalize([Type=Int, ]x; tol=eps(x)) Approximate floating point number x as a
> Rational number with components of the given integer type. The result will
> differ from x by no more than tol.
> 
> meanwhile, here is a "long way around" that may do what you want (signature and
> special cases are cribbed from rational.jl)
> 
> ```julia
> function rationalized{T<:Integer}(::Type{T}, x::AbstractFloat; tol::Real=eps(x))
>     tol < 0 && throw(ArgumentError("negative tolerance'))
>     isnan(x) && return zero(T)//zero(T)
>     isinf(x) && return (x < 0 ? -one(T) : one(T)) // zero(T)
>     rationalize(round(x, floor(Int,abs(log2(tol)))))
> end
> ```

Should the workaround be adopted as the correct version of `rationalize`?
Is this a regression that used to work correctly?

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #16311

### CODE REVIEW

File: base/rational.jl [hunk 1]
```diff
@@ -125,16 +128,21 @@ function rationalize{T<:Integer}(::Type{T}, x::AbstractFloat; tol::Real=eps(x))
...
+        # maintain
+        # x0 = (p + (-1)^i * r) / q
+        t, tt = nt, t
+        nt = a*t+tt
     end
 
     # find optimal semiconvergent
     # smallest a such that x-a*y < a*t+tt
-    a = cld(x-tt,y+t)
+    a = max(cld(x-tt,y+t),0.0)
```
Comment (tkelman) [inline base/rational.jl]: should this use `zero` for type stability?

Comment (tkelman) [inline base/rational.jl]: @vtjnash apparently thinks not...

Comment (tkelman) [inline base/rational.jl]: Simon? If we were to use `zero` here, which term's type should it use?

Comment (simonbyrne) [inline base/rational.jl]: I fixed this in #16512

**Pull request merged**